components/cloud/aws/src/s3.rs

```diff
@@ -626,6 +640,68 @@ impl BlobStorage for S3Storage {
     }
 }
 
+struct S3Walker<'cli, 'arg> {
+    cli: &'cli S3Storage,
+    finished: bool,
+    cont_token: Option<String>,
+    prefix: &'arg str,
+}
+
+impl<'cli, 'arg> S3Walker<'cli, 'arg> {
+    async fn one_page(&mut self) -> io::Result<Option<Vec<BlobObject>>> {
+        if self.finished {
+            return io::Result::Ok(None);
+        }
+        let mut input = ListObjectsV2Request::default();
+        input.bucket = String::clone(&self.cli.config.bucket.bucket);
+        input.prefix = Some(self.cli.maybe_prefix_key(self.prefix));
+        input.max_keys = Some(128);
+        input.continuation_token = self.cont_token.clone();
+        let res = self
+            .cli
+            .client
+            .list_objects_v2(input)
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
```

Comment:
Sadly it seems for now we don't have a principle guides us converting an `RusotoError` to an `IoError`... Perhaps all we can do is retry internally. Thankfully this won't be called from a remote client for now.

Reply:
let me see if I can figure out something, the standard io error type is not very convenient for error handling. Ideally every error we (TiKV) return should be a custom error type that we created. It can wrap the actual error from upstream, so we can know what's the exact place that throws the error and possibly handle it more gracefully.